### PR TITLE
Adding implementations for Hybrid-CBMs, ProbCBMs, and Post-hoc CBMs

### PR DESCRIPTION
**Pull request by mateoespinosa**
This PR adds the following features:
1. An implementation of [Hybrid CBMs](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/torch_concepts/nn/modules/high/models/hybrid_cbm.py), as defined by [Mahinpei et al. (ICML Workshop 2021)](https://arxiv.org/abs/2106.13314). This implementation is accompanied by a [script](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/examples/utilization/2_model/15_hybridcbm_torch_training.py) describing how to train Hybrid-CBMs using pure PyTorch.
2. An implementation of [Probabilistic CBMs (ProbCBMs)](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/torch_concepts/nn/modules/high/models/prob_cbm.py), as introduced by [Kim et al. (ICML 2023)](https://arxiv.org/abs/2306.01574). This implementation is accompanied by two example scripts. The [first script](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/examples/utilization/2_model/16_probcbm_torch_training.py) describes how to jointly train ProbCBMs using pure PyTorch (without any warmup or sequential training). The [second script](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/examples/utilization/2_model/16.5_probcbm_lightning_training.py) provides an example of training sequential ProbCBMs using the authors' suggested training pipeline, which has been incorporated into the Lightning instantiation of the model class.
3. An [anchor-based predictor](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/torch_concepts/nn/modules/low/predictors/anchor.py) which can be used to model the predictor used by prototype-based approaches like ProbCBMs.
4. An implementation of [Post-hoc CBMs (PCBMs)](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/torch_concepts/nn/modules/high/models/post_hoc_cbm.py), both its "standard" and its "residual" variants, as introduced by [Yuksekgonul et al. (ICLR 2023)](https://arxiv.org/abs/2205.15480). This implementation is accompanied by two example scripts. The [first script](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/examples/utilization/2_model/17_posthoccbm_torch_training.py) describes how to jointly train PCBMs using pure PyTorch. The [second script](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/examples/utilization/2_model/17.5_posthoccbm_lightning_training.py) shows how to train PCBMs using the authors' suggested training pipeline, which has been incorporated into the Lightning instantiation of the model class. For this, I made a small update to the [CAV](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/torch_concepts/nn/modules/low/encoders/cav.py) class to improve compatibility with PCBMs.
5. A very simple [residual predictor](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/torch_concepts/nn/modules/low/predictors/residual.py) which can be used to fit a residual channel into an already interpretable predictor (if this is not that useful, I am happy to fold it into the PCBM class itself).
6. A small change to the [learner class](https://github.com/pyc-team/pytorch_concepts/blob/mateo/dev/torch_concepts/nn/modules/high/base/learner.py) that enabled me to easily incorporate more complex training pipelines while reusing standard MLE estimation losses.
7. A set of simple unit tests to check the sanity of the new models through future updates of the repo.
8. Some very minor stylistic cleanups to some files.

### CODE REVIEW

File: torch_concepts/nn/modules/low/predictors/anchor.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+                (batch, m, embedding_size).
+        """
+        anchors = self._anchors_for(items)
+        weights = values.reshape(values.shape[0], anchors.shape[0], -1)
+        if self.cardinality == 1:
+            weights = torch.cat([1.0 - weights, weights], dim=-1)
+        return (weights.unsqueeze(-1) * anchors.unsqueeze(0)).sum(-2)
+
+
+class AnchorPredictor(BaseConceptLayer):
```
Comment (gdefe) [inline torch_concepts/nn/modules/low/predictors/anchor.py:248]: We recently dropped the distinction between Encoders and Predictors. They are all ConceptLayers (they all output concepts, not worth distinguishing based on minor changes in the input signature)

If this is a concpet layer, we should use their naming convention:

[Layer][Input]To[Output]

File: torch_concepts/nn/modules/low/predictors/anchor.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+        torch.Size([8, 4])
+
+    References:
+        Kim et al. "Probabilistic Concept Bottleneck Models", ICML 2023.
+        https://arxiv.org/abs/2306.01574
+    """
+
+    def __init__(
+        self,
+        n_items: int,
```
Comment (gdefe) [inline torch_concepts/nn/modules/low/predictors/anchor.py:308]: It would be great if we can somehow uniform the signature of this with the other concept layers.
Meaning we should have:

in_concepts
in_embeddings
out_concepts

Do you think is possible?

An important point that could help: we can force 'in_concepts' to be of Annotations type (see for example the mixing cem layer at MixConceptEmbeddingToConcept), this would give access to concept cardinalities at construction without the extra param

File: torch_concepts/nn/__init__.py [hunk 3]
```diff
@@ -59,6 +64,9 @@
 # Models (high-level)
 from .modules.high.models.blackbox import BlackBox, BlackBoxTaskOnly
 from .modules.high.models.cbm import ConceptBottleneckModel
+from .modules.high.models.hybrid_cbm import HybridConceptBottleneckModel
```
Comment (gdefe) [inline torch_concepts/nn/__init__.py:67]: Should we standardize high-level names as well? I am not sure.

File: torch_concepts/nn/modules/high/models/hybrid_cbm.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+    unsupervised_distributions = {
+        'binary': Bernoulli,
+        'continuous': Delta,
+    }
+
+    def __init__(
+        self,
+        input_size: int,
+        annotations: Annotations,
+        additional_dims: int,
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/models/hybrid_cbm.py:151]: shall we term it "embedding_size", so it is uniform with the other models?

File: torch_concepts/nn/modules/high/models/hybrid_cbm.py [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+        # latent -> unsupervised dimensions
+        unsup_encoders = ParametricCPD(
+            variable=unsup,
+            parents=[latent_var],
+            parametrization=[
+                self._flexible_parametrization(
+                    variable=u,
+                    first=LazyConstructor(LinearEmbeddingToConcept),
+                    second='auto',
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/models/hybrid_cbm.py:345]: we switched from 'auto' -> 'copy', same functionality, just a change in the name

File: torch_concepts/nn/modules/high/base/learner.py [hunk 6]
```diff
@@ -296,6 +297,39 @@ def default_evidence(self, inputs):
         """
         return {"input": inputs["x"]}
 
+    def supervised_loss(self, out, target, names):
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/base/learner.py:300]: For what this method computes, ConceptSubset + ConceptLoss should already cover it, and it should be faster and more robust.

For example, if I want to compute a BCE only on a subset of binary concepts, I would do:

loss = ConceptSubset(ConceptLoss(binary=BCEWithLogitsLoss()), names=[..., ...., ....])

Should we give it a try? example 2_model/13_composite_loss show all of such loss combinations.

What it is missing, is a clean way to tie a specific loss to a model that come with one. Any idea? 
1) Maybe a ProbCBMLoss class in the same model file, which internally build the loss with all the necessary haparams?
2) Maybe we do not hardocode anything in the py file, but specifcy each model default in the yaml? including classes? This scales better, but forces people to understand hydra conf

My bad, things should be better documented :(

File: examples/utilization/2_model/15_hybridcbm_torch_training.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+# concept values, so a noiseless input determines every concept almost exactly
+# (~100% concept accuracy) and leaves an expert nothing to correct. Degrading
+# the labels will help us make concept interventions a bit more meaningful in
+# the examples shown below.
+INPUT_NOISE = 0.75
+
+# Rate at which training replaces a predicted concept by its ground truth
+# ("RandInt"): the task head then also learns from hard concept values, which
+# is what keeps a model responsive to interventions at test time.
+P_INT = 0.0
```
Comment (gdefe) [inline examples/utilization/2_model/15_hybridcbm_torch_training.py:57]: I would set this to a higher default

File: examples/utilization/2_model/17.5_posthoccbm_lightning_training.py
Comment (gdefe) [inline examples/utilization/2_model/17.5_posthoccbm_lightning_training.py:1]: I would name files 16.1 and 16.2, same for 17.1 and 17.2. This is the convention we are usinge elsewhere for multiple related examples

File: torch_concepts/nn/modules/high/models/prob_cbm.py [hunk 8]
```diff
@@ -0,0 +1,797 @@
...
+                block[:, position] = anchored[:, present.index(member)]
+            # Reshape to the variable's own event shape, which is what teacher
+            # forcing compares against (evidence would accept either).
+            evidence[name] = block.reshape(
+                block.shape[0],
+                *self.pgm.variables[name].shape,
+            )
+        return evidence
+
+    def vib_kl(self, out: ModelOutput) -> torch.Tensor:
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/models/prob_cbm.py:442]: we have implemented a KL loss in another PR for generative models. After I merge that, we can maybe use it for this?

File: torch_concepts/nn/modules/high/models/post_hoc_cbm.py [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+
+    def configure_optimizers(self):
+        """
+        Adam over whatever the current stage left trainable, as in the paper.
+
+        No schedule: the reference fits the interpretable head with a sparse
+        solver and the residual with a plain constant-rate Adam, so there is
+        nothing to anneal.
+        """
+        return dict(
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/models/post_hoc_cbm.py:464]: The BaseLearner class already allows to pass your custom scheduler. We should find a robust and scalable way of linking a proposed model to its loss and optimizer. See the other comment in the learner.py

File: torch_concepts/nn/modules/high/models/prob_cbm.py [hunk 10]
```diff
@@ -0,0 +1,797 @@
...
+        task_names: Union[List[str], str],
+        embedding_size: int = 16,
+        class_embedding_size: int = 128,
+        init_negative_scale: float = 5.0,
+        init_class_scale: float = 5.0,
+        per_item_scale: bool = False,
+        use_anchor_interpolation: bool = False,
+        # -- Training recipe: only read when lightning=True ---------------
+        vib_beta: float = 5e-5,
+        intervention_prob: float = 0.5,
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/models/prob_cbm.py:203]: shouldnt this be a parameter of the inference?

File: torch_concepts/nn/modules/high/models/prob_cbm.py [hunk 11]
```diff
@@ -0,0 +1,797 @@
...
+        train_class_mode: str = 'sequential',
+        concept_epochs: int = 50,
+        class_epochs: int = 20,
+        warm_epochs: int = 5,
+        lr: float = 1e-3,
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/models/prob_cbm.py:208]: In the other models, these are passed via "optim_kwargs" to instantiate internally an optimizer class passed via 'optim_class'. 

I suppose here you made this explicit to bind the paper's implementation, but I hope we could find a standard way to do this.
